pkg/operator/controllers/azurensg/subnetnsg_test.go

```diff
@@ -0,0 +1,71 @@
+package azurensg
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	"github.com/Azure/go-autorest/autorest/azure"
+	"github.com/Azure/go-autorest/autorest/to"
+	"github.com/golang/mock/gomock"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	utillog "github.com/Azure/ARO-RP/pkg/util/log"
+	mock_network "github.com/Azure/ARO-RP/pkg/util/mocks/azureclient/mgmt/network"
+)
+
+var (
+	vnet = azure.Resource{
+		SubscriptionID: "0000000-0000-0000-0000-000000000000",
+		ResourceGroup:  "vnet-rg",
+		ResourceName:   "vnet-name",
+	}
+)
+
+func TestFixSubnet(t *testing.T) {
+	r := AzureNSGReconciler{log: utillog.GetLogger()}
+	for _, tt := range []struct {
+		nsgname             string
+		architectureVersion api.ArchitectureVersion
+		expectedValue       string
+	}{
+		{
+			nsgname:             "someotherv1-nsg",
+			architectureVersion: api.ArchitectureVersionV1,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-node-nsg",
+		},
+		{
+			nsgname:             "infraid-nsg",
+			architectureVersion: api.ArchitectureVersionV2,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-nsg",
+		},
+		{
+			nsgname:             "someotherv2-nsg",
+			architectureVersion: api.ArchitectureVersionV2,
+			expectedValue:       "/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/infraid-nsg",
+		},
+	} {
+		controller := gomock.NewController(t)
+		fakeSubnet := mgmtnetwork.Subnet{SubnetPropertiesFormat: &mgmtnetwork.SubnetPropertiesFormat{
+			NetworkSecurityGroup: &mgmtnetwork.SecurityGroup{
+				ID: to.StringPtr("/subscriptions/0000000-0000-0000-0000-000000000000/resourceGroups/resourcegroup/providers/Microsoft.Network/networkSecurityGroups/" + tt.nsgname),
+			},
+		}}
+		subnetsClient := mock_network.NewMockSubnetsClient(controller)
+		subnetsClient.EXPECT().Get(context.Background(), "vnet-rg", "vnet-name", "subnet", "").
+			Return(fakeSubnet, nil)
+		subnetsClient.EXPECT().CreateOrUpdateAndWait(context.Background(), "vnet-rg", "vnet-name", "subnet", fakeSubnet).
+			Return(nil)
+		err := r.fixSubnet(context.Background(), tt.architectureVersion, vnet.SubscriptionID, vnet.ResourceName, "subnet", vnet.ResourceGroup, "resourcegroup", "infraid", true, subnetsClient)
```

Comment:
With changes suggested [here](https://github.com/Azure/ARO-RP/pull/1424/files#r629925761), I think we should test via `reconcileSubnetNSG`: this will give us more coverage. Currently we do not test logic in `reconcileSubnetNSG`.